### ISSUE TITLE
P12.6: Measure data scaling, exact-score preference, and large-case qualification

### ISSUE THREAD

**Issue by kiwi1005** [kiwi1005/CadC26#30]
Parent: #24

Depends on: #26 KEEP; #27/#28/#29 only when each is individually KEEP

Branch: [`research/model-first-e2e-floorplanner`](https://github.com/kiwi1005/CadC26/tree/research/model-first-e2e-floorplanner)

# Hypothesis

Once direct model generation has source-held-out QoR value, more FloorSet training data and exact-score preference learning can improve full-program quality, especially on 106–120 block cases, without reintroducing a traditional global solver.

# Scope rule

This issue is conditional. Do not begin data scaling or preference optimization because the dataset contains 1M layouts. Begin only after #26 proves that model-generated programs beat canonical/random one-shot topology.

# Phase A: data scaling curve

Freeze:

- model architecture;
- compiler;
- source-held-out evaluation bucket;
- K and inference rounds;
- loss weights initially.

Train with nested source sets:

```text
10K
50K
200K
1M only if the curve is still rising
```

Report separately:

```text
40–80
81–105
106–120 blocks
```

Metrics:

- compile success;
- hard feasibility;
- official-style QoR;
- area/HPWL and B/G/M debt;
- unique wins;
- K-sample diversity;
- training cost and inference cost.

Stop 1M if 10K→50K has no material decoded-QoR gain.

# Phase B: large-case specialization

Only after general scaling evidence:

- score-aware sampling for 106–120 blocks;
- optional large-case sample token or specialist head;
- keep one shared encoder unless a small specialist ablation has clear unique wins;
- no case-ID routing.

Evaluate internal source-held-out large bucket, then large15/full100 only after the model and hyperparameters are frozen.

# Phase C: exact-score preference

Generate K model programs per training case and compile/score every program exactly.

Use one minimal preference objective:

```text
listwise ranking of model programs by exact official-style outcome
```

Only if listwise preference fails for a known reason may offline RL be considered. Do not start with PPO or a general RL framework.

Preference supervision may use:

- hard-feasibility tier;
- uncapped cost;
- cap margin;
- B/G/M debt;
- HPWL/bbox tie-breaks.

The model still generates all topology; exact scoring only supplies learning signal.

# Phase D: contest qualification

After model/data/preference choices are frozen:

```text
medium held-out
large held-out
large15
full100
repeated runtime
CUDA/A100 parity
submission smoke
```

Runtime path:

```text
model K programs
-> exact compiler K
-> optional bounded model-led residual from #29
-> best verified candidate
-> safety fallback only if every model path fails
```

# Promotion metrics

- full100 hard feasibility 100/100 final output;
- model compile success and fallback rate reported separately;
- weighted QoR improvement over the current direct-generation checkpoint;
- large-case weighted gain / cap crossings;
- percentage of final winners originating from model programs;
- K/round/runtime trade-off;
- residual algorithm contribution bounded and explicit.

# Decision

## KEEP scaling when

- decoded QoR improves monotonically or meaningfully with data;
- large-case gain appears on source-held-out cases;
- model-generated programs remain the source of the gain.

## KEEP preference when

- exact-score preference improves decoded QoR, cap margin or winner rate over supervised-only training;
- hard feasibility/compile success are preserved;
- no solver trajectory is required as model input.

## REJECT / stop when

- 10K→50K is saturated;
- only training losses scale;
- large gains require case-specific routing;
- exact preference merely teaches fallback selection;
- algorithmic cleanup supplies most of the improvement;
- model-first path is slower without meaningful QoR gain.

# Deliverables

- frozen nested source manifests;
- scaling report and curve;
- exact preference replay format only if Phase A KEEP;
- medium/large/full100 report;
- model/algorithm contribution accounting;
- final KEEP/MODIFY/REJECT decision.

# Non-goals

- no generic RL platform;
- no new global search family;
- no scaling before the direct-generation gate;
- no release engineering until model-first QoR is proven.
